DRILL-5783 Made a unit test for generated Priority Queue. DRILL-5841 Fix tmp folder errors. DRILL-5894 clean up usage of dfs_test in tests.

ilooner opened the pull request:
## DRILL-5783

- A unit test is created for the priority queue in the TopN operator
- The code generation classes passed around a completely unused function registry reference in some places so I removed it.
- The priority queue had unused parameters for some of its methods so I removed them.

## DRILL-5841

- There were many many ways in which temporary folders were created in unit tests. I have unified the way these folders are created with the DirTestWatcher, SubDirTestWatcher, and BaseDirTestWatcher. All the unit tests have been updated to use these. The test watchers create temp directories in ./target/<fully qualified test class name>/<method name>. So all the files generated and used in the context of a test can easily be found in the same consistent location.
- This change should fix the sporadic hashagg test failures, as well as failures caused by stray files in /tmp

## DRILL-5894

- **dfs_test** is used as a storage plugin throughout the unit tests. This is highly confusing and we can just use *dfs* instead.

Misc

- General code cleanup. 
- There are many places where String.format is used unnecessarily. The test builder methods already use String.format for you when you pass them args. I cleaned some of these up.

ilooner commented:
@paul-rogers

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/compile/CodeCompiler.java:
```diff
@@ -110,6 +109,11 @@ public CodeGenCompiler(final DrillConfig config, final OptionSet optionManager)
   /**
+   * Enables saving generated code for debugging
+   */
+  public static final String ENABLE_SAVE_CODE_FOR_DEBUG = COMPILE_BASE + ".codegen.dump";
```
The directory that generated code is dumped into is determined by the ClassBuilder.CODE_DIR_OPTION property. The DirTestWatcher createw a directory of the following form:

```
/target/<Fully qualified test class name>/<test method name>
```

So if the ClassBuilder.CODE_DIR_OPTION property is configured to use the directory generated by the DirTestWatcher, the code is dumped into the same well defined location every test run. Also the generated class file has names like:

```
<interface name>Gen<random num>.java
```

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/expr/CodeGenerator.java:
```diff
@@ -90,12 +89,11 @@
-  CodeGenerator(TemplateClassDefinition<T> definition, FunctionImplementationRegistry funcRegistry, OptionSet optionManager) {
-    this(ClassGenerator.getDefaultMapping(), definition, funcRegistry, optionManager);
+  CodeGenerator(TemplateClassDefinition<T> definition, OptionSet optionManager) {
+    this(ClassGenerator.getDefaultMapping(), definition, optionManager);
```
I traced through the code and asked IntelliJ where the variable is used. This is what I found:

1. CodeGenerator accepts it as an argument to its constructor
1. It is then passed to the constructor of the EvaluationVisitor
1. A private field is set with the value in the constructor of EvaluationVisitor
1. This private field of EvaluationVisitor is then unused.

If your not convinced let's walk through the code when I'm in the office next week.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/PriorityQueue.java:
```diff
@@ -20,22 +20,58 @@
+  void init(int limit, BufferAllocator allocator, boolean hasSv2) throws SchemaChangeException;
+
+  /**
+   * This method must be called before fetching the final heap hyper batch and final Sv4 vector.
```
Old habit, I'm used to PriorityQueues being called Heaps. I agree it's confusing in this context I will fix it.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/compile/CodeCompiler.java:
```diff
@@ -110,6 +109,11 @@ public CodeGenCompiler(final DrillConfig config, final OptionSet optionManager)
   /**
+   * Enables saving generated code for debugging
+   */
+  public static final String ENABLE_SAVE_CODE_FOR_DEBUG = COMPILE_BASE + ".codegen.dump";
```
This is the name of the option which holds the true/false flag to enable saving code for debugging. The destination directory is governed by the ClassBuilder.CODE_DIR_OPTION property.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java:
```diff
@@ -95,7 +91,9 @@ public TopNBatch(TopN popConfig, FragmentContext context, RecordBatch incoming)
-    batchPurgeThreshold = context.getConfig().getInt(ExecConstants.BATCH_PURGE_THRESHOLD);
+    DrillConfig drillConfig = context.getConfig();
+    batchPurgeThreshold = drillConfig.getInt(ExecConstants.BATCH_PURGE_THRESHOLD);
+    codegenDump = drillConfig.getBoolean(CodeCompiler.ENABLE_SAVE_CODE_FOR_DEBUG);
```
I will rename the option to drill.debug.codegen.TopN

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java:
```diff
@@ -335,20 +333,32 @@ private void purge() throws SchemaChangeException {
+          throws ClassTransformationException, IOException, SchemaChangeException {
+    final MappingSet mainMapping = new MappingSet((String) null, null, ClassGenerator.DEFAULT_SCALAR_MAP, ClassGenerator.DEFAULT_SCALAR_MAP);
+    final MappingSet leftMapping = new MappingSet("leftIndex", null, ClassGenerator.DEFAULT_SCALAR_MAP, ClassGenerator.DEFAULT_SCALAR_MAP);
+    final MappingSet rightMapping = new MappingSet("rightIndex", null, ClassGenerator.DEFAULT_SCALAR_MAP, ClassGenerator.DEFAULT_SCALAR_MAP);
```
Reverted it back to the way it was

ilooner commented on common/src/test/java/org/apache/drill/testutils/SubDirTestWatcher.java:
```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+
+package org.apache.drill.testutils;
```
done

ilooner commented:
@paul-rogers Applied / responding to comments. Also removed RecordBatchBuilder and used RowSetBuilder.

paul-rogers commented on exec/vector/src/main/codegen/templates/FixedValueVectors.java:
```diff
@@ -908,6 +908,15 @@ public void generateTestData(int count) {
+      int valueCount = index + 1;
+      setValueCount(valueCount);
+
+      set(index, (${type.javaType}) value);
```
I don't think this will work as you hope it will -- for obscure reasons that we've all learned the hard way...

* The accessor does not keep an ongoing count of values. Rather, the value count comes from the writer index which is set only at the end of a write.
* This method sets the value count, but doing so is, in general, expensive.
* Vectors in a batch must be kept in sync. This adds a value to one vector, but does not coordinate across a set of vectors.
* Not all values can be set by casting. Works for some numeric values, but does not handle, say, `add(10)` when the item is a byte or a short because of the need for a two-step cast: `(byte)(int) value`.
* Does not handle conversions for date or decimal.

Yes, this is all a colossal pain in the neck. But, it is the reason that the `RowSet` classes were created.

We can talk in person about how to move this idea forward effectively.

ilooner replied:
All the issues with this make sense. I should have deleted this when I switched to using the RowSet classes, but forgot too :). Removing this now.

ilooner-mapr commented on common/src/main/java/org/apache/drill/common/util/TestTools.java:
```diff
@@ -17,15 +17,28 @@
 public class TestTools {
   // private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(TestTools.class);
 
+  public enum DataType {
```
Done changed to FileSource

julianhyde commented on exec/java-exec/src/test/java/org/apache/drill/test/ClusterFixture.java:
```diff
@@ -584,11 +492,14 @@ public static void defineWorkspace(Drillbit drillbit, String pluginName,
 
-  public static FixtureBuilder builder() {
-    FixtureBuilder builder = new FixtureBuilder()
+  public static FixtureBuilder builder(DirTestWatcher dirTestWatcher) {
```
Jeez Paul, please start a review rather than making single review comments. I just got 39 emails from you, and so did everyone else on dev@drill.

ilooner-mapr commented on common/src/test/java/org/apache/drill/test/DirTestWatcher.java:
```diff
@@ -32,23 +32,50 @@
 public class DirTestWatcher extends TestWatcher {
   private String dirPath;
   private File dir;
+  private boolean deleteDirAtEnd = true;
```
Done

ilooner-mapr commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/PriorityQueue.java:
```diff
@@ -20,22 +20,58 @@
-  public void cleanup();
-
-  public static TemplateClassDefinition<PriorityQueue> TEMPLATE_DEFINITION = new TemplateClassDefinition<PriorityQueue>(PriorityQueue.class, PriorityQueueTemplate.class);
+  /**
```
I think the priority queue is generic and not specific to TopN. I agree we should make an effort to consolidate the code generated data structures.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/PriorityQueue.java:
```diff
@@ -20,22 +20,58 @@
+   * queue are called.
+   * @param limit The size of the priority queue.
+   * @param allocator The {@link BufferAllocator} to use when creating the priority queue.
+   * @param hasSv2 True when incoming batches have v2 selection vectors. False otherwise.
```
fixed

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/PriorityQueue.java:
```diff
@@ -20,22 +20,58 @@
+   * provided by the {@link #getHyperBatch()} method. <b>Note:</b> this should be called after {@link #generate()}.
+   * @return The selection vector used to select the elements in the priority queue.
+   */
+  SelectionVector4 getFinalSv4();
```
The code already works that way. There is a config option <b>drill.exec.sort.purge.threshold</b> which controls the maximum number of batches allowed in the hyper batch. Once the threshold is exceeded the top N values are consolidated into a single batch and the process is repeated. There is an issue in the case where the limit is large. Ex. 100,000,000 . In this case the operator will keep all the records in memory and die. There is a jira created to address this issue: [https://issues.apache.org/jira/browse/DRILL-5823]

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java:
```diff
@@ -335,20 +336,42 @@ private void purge() throws SchemaChangeException {
+  }
+
+  public static PriorityQueue createNewPriorityQueue(
+    MappingSet mainMapping, MappingSet leftMapping, MappingSet rightMapping,
```
Agreed. Pritesh has asked me to look at Codegen improvements in the long term, so I will work on this down the line.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector4.java:
```diff
@@ -116,6 +116,14 @@ public void clear() {
+
+  public static int getRecordIndex(int sv4Index) {
+    return (sv4Index) & 0xFFFF;
+  }
```
Didn't do any performance testing. I mainly intended to use these methods for testing purposes. For example I use them in org.apache.drill.test.BatchUtils .

ilooner commented on exec/java-exec/src/main/resources/drill-module.conf:
```diff
@@ -247,7 +248,7 @@ drill.exec: {
   },
   sort: {
-    purge.threshold : 10,
+    purge.threshold : 1000,
```
Oh my. Thanks for catching that!

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/TestAggNullable.java:
```diff
@@ -21,29 +21,25 @@
-    test(String.format("alter session set `planner.enable_hashagg` = %s", ha ? "true":"false"));
-    test(String.format("alter session set `planner.enable_streamagg` = %s", sa ? "true":"false"));
+    test("alter session set `planner.enable_hashagg` = %s", ha);
+    test("alter session set `planner.enable_streamagg` = %s", sa);
```
We can update these after the other PR goes in then.

ilooner commented on exec/java-exec/src/test/resources/topN/one_key_sort.json:
```diff
@@ -12,11 +12,11 @@
             url: "http://apache.org",
             entries:[
-                {records: 10000000, types: [
+                {records: 100000, types: [
```
The original test was checking to make sure the the Top N results returned are ordered. It was never really necessary to generate that much data. The benefit of reducing the amount of data generated is that the test runs faster.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/TestAltSortQueries.java:
```diff
@@ -64,9 +73,9 @@ public void testJoinWithLimit() throws Exception{
         "  regions.R_NAME\n" +
         "FROM\n" +
-        "  dfs_test.`[WORKING_PATH]/../../sample-data/nation.parquet` nations\n" +
+        "  dfs.`/sample-data/nation.parquet` nations\n" +
```
I accidentally changed it to dfs, will switch it back to dfs_test for consistency. <b>dfs</b> and <b>dfs_test>/b> are configured to point to the same temp directories so they are in fact interchangeable. This configuration happens in the openClient method of BaseTestQuery and in the configureStoragePlugins method of the ClusterFixture. 

Throughout making this change I was scratching my head as to why <b>dfs_test</b> is used instead of <b>dfs<b>. After going through the code I don't see any reason as to why dfs_test was created in the first place. I think we should actually remove <b>dfs_test</b> and just use <b>dfs</b> everywhere for uniformity. I'll go ahead and do that now.

ilooner replied:
Just mentioned this above but will repeat here.I have now removed **dfs_test** completely. There was no reason for it to be added and it was inconsistently being mixed with **dfs**. If you want to query a file on the local filesystem that is not on the classpath just using **dfs** will be sufficient now.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/TestAltSortQueries.java:
```diff
@@ -19,24 +19,33 @@
   public void testOrderBy() throws Exception{
     test("select R_REGIONKEY " +
-         "from dfs_test.`[WORKING_PATH]/../../sample-data/region.parquet` " +
+         "from dfs_test.`/sample-data/region.parquet` " +
```
I have now removed **dfs_test** completely. There was no reason for it to be added and it was inconsistently being mixed with **dfs**. The **dfs** workspaces are automatically mapped to the correct temp directories for you provided that you use **BaseTestQuery** or the **ClusterFixture**. I will update **org.apache.drill.test.package-info.java** with the theory of how this works and will add a simple example to **ExampleTest.java**

ilooner replied:
Updated package-info.java and ExampleTest.java

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/TestCTASPartitionFilter.java:
```diff
@@ -59,48 +58,48 @@ public void withDistribution() throws Exception {
+    test("create table orders_distribution partition by (o_orderpriority) as select * from dfs_test.`/multilevel/parquet`");
     String query = "select * from orders_distribution where o_orderpriority = '1-URGENT'";
-    testExcludeFilter(query, 1, "Filter", 24);
+    testExcludeFilter(query, 1, "Filter\\(", 24);
```
It is no longer sufficient to match "Filter" because the test class name contains "Filter" and the test class name is used to create the tmp directory. And the fully qualified path of a queried file is included in the plan. We want to only match the <b>Filter</b> steps generated in the plan, not the <b>Filters</b> in our file paths. In order to do this I tell it to match "Filter(" which corresponds to a filter step in the plan.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/TopN/TopNBatchTest.java:
```diff
@@ -0,0 +1,179 @@
+          .container();
+        hyperContainer = new ExpandableHyperContainer(container);
+
+        queue = TopNBatch.createNewPriorityQueue(
```
Agreed. Sounds good.

ilooner commented on common/src/test/java/org/apache/drill/test/SubDirTestWatcher.java:
```diff
@@ -0,0 +1,108 @@
+import java.io.File;
+import java.util.List;
+
+public class SubDirTestWatcher extends TestWatcher {
```
Done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/TestBugFixes.java:
```diff
@@ -33,8 +35,11 @@
+
+  @BeforeClass
+  public static void setupTestFiles() {
+    dirTestWatcher.copyResourceToRoot("/bugs/DRILL-4192");
```
I have made all the path references relative in the tests now

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestCorruptParquetDateCorrection.java:
```diff
@@ -76,43 +78,50 @@
+  private static final String PARTITIONED_1_2_FOLDER = "partitioned_with_corruption_4203_1_2";
   private static final String CORRUPTED_PARTITIONED_DATES_1_2_PATH =
-      "[WORKING_PATH]/src/test/resources/parquet/4203_corrupt_dates/partitioned_with_corruption_4203_1_2";
+      Paths.get("/parquet/4203_corrupt_dates", PARTITIONED_1_2_FOLDER).toString();
```
Updated all paths to be relative in the tests

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestCorruptParquetDateCorrection.java:
```diff
@@ -76,43 +78,50 @@
-
   @BeforeClass
   public static void initFs() throws Exception {
+    dirTestWatcher.copyResourceToRoot("/parquet/4203_corrupt_dates");
```
fixed

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/xsort/TestSimpleExternalSort.java:
```diff
@@ -147,7 +152,7 @@ public void sortOneKeyDescendingExternalSortLegacy() throws Throwable {
 
   private void sortOneKeyDescendingExternalSort(boolean testLegacy) throws Throwable {
-    FixtureBuilder builder = ClusterFixture.builder()
+    FixtureBuilder builder = ClusterFixture.builder(dirTestWatcher)
```
Is there a case where the tmp directories are optional? The Drillbit will always have to be configured with a <b>drill.tmp-dir</b> at minimum, so I think it should be required to use a BaseDirTestWatcher with a ClusterFixture. I could move passing the dirTestWatcher to the build method instead of the constructor. Let me know what you think.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BaseDirTestWatcher.java:
```diff
@@ -0,0 +1,184 @@
+    }
+  }
+
+  public File makeRootSubDir(String relPath) {
```
Fair enough. I would prefer to use File objects to represent concrete files or directories. Java.nio.Path objects are useful because they can represent relative paths, and provide a bunch of methods for manipulating paths. I'll update the methods I introduced to use jva.io.Files and java.nio.Paths

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/ClusterFixture.java:
```diff
@@ -213,7 +204,7 @@ private void configureZk(FixtureBuilder builder) {
   }
 
-  private void createConfig(FixtureBuilder builder) throws Exception {
+  private void createConfig() throws Exception {
```
They should be. The only change I made is that I didn't require passing the FixtureBuilder as an arg, because the class is already holding a reference to it.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/ClusterFixture.java:
```diff
@@ -226,43 +217,16 @@ private void createConfig(FixtureBuilder builder) throws Exception {
-    removeLocalFiles();
+  private void startDrillbits() throws Exception {
+    dfsTestTempDir = new File(getRootDir(), "dfs-test");
+    dfsTestTempDir.mkdirs();
```
This is possible through the BaseDirTestWatcher. It is configured through the BaseDireTestWatcher's constructor. If you pass it false it will not delete your directories at the end of each test.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/ExampleTest.java:
```diff
@@ -59,6 +59,9 @@
 public class ExampleTest {
 
+  @Rule
+  public final DirTestWatcher dirTestWatcher = new DirTestWatcher();
```
I added java doc, and added example of how it's used in secondTest

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BaseDirTestWatcher.java:
```diff
@@ -0,0 +1,184 @@
+    }
+  }
+
+  public void replaceMetaDataContents(File metaDataFile, String replacePath, String customStringReplacement) {
```
done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BatchUtils.java:
```diff
@@ -0,0 +1,280 @@
+    return rows;
+  }
+
+  public static Map<String, List<Object>> hyperBatchAndSelectorToObjects(VectorContainer vectorContainer, SelectionVector4 selectionVector4) {
```
Done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BatchUtils.java:
```diff
@@ -0,0 +1,280 @@
+    return sb.toString();
+  }
+
+  public static void assertEqual(Map<String, List<Object>> expected, Map<String, List<Object>> actual) {
```
Done

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BatchUtils.java:
```diff
@@ -0,0 +1,280 @@
+    }
+  }
+
+  public static class ObjectComparator implements Comparator<Object> {
```
I looked at RowSetComparison. It had code for testing equality, but not for ordering (less than, greater than) which is what I needed for my test. I've moved this class into RowSetComparison without any changes for now. We need to modify DrillTestWrapper to use the RowSet classes first, then we can make ObjectComparator to be something native to RowSets.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/BatchUtils.java:
```diff
@@ -0,0 +1,280 @@
+public class BatchUtils {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(BatchUtils.class);
+
+  public static Map<String, List<Object>> containerToObjects(VectorContainer vectorContainer) {
```
Removed BatchUtils

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/xsort/TestExternalSort.java:
```diff
@@ -138,34 +141,34 @@ public void testNewColumnsManaged() throws Exception {
-      }
+    final String tableDirName = "newColumns";
+
+    TableFileBuilder tableA = new TableFileBuilder(Lists.newArrayList("a", "b"), Lists.newArrayList("%d", "%d"));
```
I've made the JsonFileBuilder, which takes a RowSet and writes it out to a file as json.

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestCorruptParquetDateCorrection.java:
```diff
@@ -377,21 +386,21 @@ public void testReadOldMetadataCacheFileOverrideCorrection() throws Exception {
+    File meta = dirTestWatcher.copyResourceToRoot(
         "parquet/4203_corrupt_dates/mixed_version_partitioned_metadata.requires_replace.txt",
-        MIXED_CORRUPTED_AND_CORRECT_PARTITIONED_FOLDER, Metadata.METADATA_FILENAME);
+        Paths.get(MIXED_CORRUPTED_AND_CORRECT_PARTITIONED_FOLDER, Metadata.METADATA_FILENAME).toString());
```
I've cleaned this up a bit. Of the two patterns for building paths and files I prefer pattern **A** because it is cleaner.

**Pattern A:**
```
myPath.resolve("subDir1")
  .resolve("subDir2")
  .toFile();
```

**Pattern B:**
```
new File(new File(myFile, "subDir1"), "subDir2")
```

ilooner commented on exec/java-exec/src/test/java/org/apache/drill/test/TableFileBuilder.java:
```diff
@@ -0,0 +1,85 @@
+  private final List<Object[]> rows = Lists.newArrayList();
+  private final String rowString;
+
+  public TableFileBuilder(List<String> columnNames, List<String> columnFormatters) {
```
I've switched this around a bit. The general flow is now the following:

  1. A **RowSet** is constructed with a **RowSetBuilder**
  1. A **JsonFileBuilder** is created and configured to use the **RowSet** that was just created.
  1. The **JsonFileBuilder** reads the **RowSet** and writes it out to a file in a json format.

ilooner commented:
@paul-rogers This is ready for another round of review. I have also responded to / addressed your comments.

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/test/rowSet/file/JsonFileBuilder.java:
```diff
@@ -0,0 +1,159 @@
+    return this;
+  }
+
+  public void build(File tableFile) throws IOException {
```
Great! This does not yet handle nested tuples or arrays; in part because the row set work for that is still sitting in PR #914. You can update this to be aware of maps and map arrays once that PR is committed.

ilooner replied:
Sounds Good

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/test/rowSet/RowSetComparison.java:
```diff
@@ -255,4 +257,39 @@ private void verifyArray(String colLabel, ArrayReader ea,
   }
+
+  // TODO make a native RowSetComparison comparator
+  public static class ObjectComparator implements Comparator<Object> {
```
Defined here, but not used in this file. Does not include all types that Drill supports (via the RowSet): Date, byte arrays, BigDecimal, etc. Does not allow for ranges for floats & doubles as does JUnit. (Two floats are seldom exactly equal.)

ilooner replied:
This is used in the DrillTestWrapper to verify the ordering of results. I agree this is not suitable for equality tests, but it's intended to be used only for ordering tests. I didn't add support for all the supported RowSet types because we would first have to move DrillTestWrapper to use RowSets (currently it uses Maps and Lists to represent data). Currently it is not used by RowSets, but the intention is to move DrillTestWrapper to use RowSets and then make this comparator operate on RowSets, but that will be an incremental process.

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/test/rowSet/RowSet.java:
```diff
@@ -85,8 +85,7 @@
    */
-
-  public interface RowSetWriter extends TupleWriter {
+  interface RowSetWriter extends TupleWriter {
```
Aren't nested interfaces `protected` by default? Just had to change one from default to `public` so I could use it in another package...

ilooner replied:
IntelliJ gave a warning that the modifier is redundant. Also an interface nested inside another interface is public by default.

https://beginnersbook.com/2016/03/nested-or-inner-interfaces-in-java/

paul-rogers replied:
Ah, forgot that the file defines an interface, not a class. (The situation I described was an interface nested inside a class.) So, you're good.

ilooner commented:
After squashing the commits and rebasing I noticed the windows functional tests were failing. The issue was caused by replacing the '/' constant in FileUtils (now renamed to DrillFileUtils) in **ClassTransformer** and **FunctionInitializer** with File.separator. This broke the build because File.separator is '\' on windows and in the context of **ClassTransformer** and **FunctionInitializer** the separator is used to look things up in the classpath. Looking things up in the classpath requires '/' on both windows and linux, hence I added back the '/' constant to DrillFileUtils along with a comment explaining its purpose and used it in **ClassTransformer** and **FunctionInitializer**. *Note:* this was a minor fix that impacted only a few lines.

arina-ielchiieva commented:
@ilooner changes in 361 files seems to be too much :) I suggest you split this PR at least at three according to the addressed issues. So we try to merge smaller chunks.

ilooner commented:
@arina-ielchiieva I can put **DRILL-5783** in a separate PR **DRILL-5841** and **DRILL-5894** are however tightly coupled and must be in the same PR. Unfortunately **DRILL-5841** and **DRILL-5894** touched many files because all the unit tests were updated so the PR will still remain large.

arina-ielchiieva commented:
@ilooner as far as know you are doing presentation today connected with this pull request. If overall approach will be accepted, then we'll try to merge this pull request. Please resolve the conflicts as well.

ilooner commented:
Resolved conflicts.

paul-rogers commented:
Discussed in person. Conflicts are resolved. We have a plan for merging with other in-flight PRs.
+1 (again)
